### PR TITLE
Feature/tagged examples

### PR DESCRIPTION
**Pull request by tencek**
Please have a look at `FeatureParser` + `BlockParser` which implements the tagged examples feature.
The rest of the changes is just to allow me to build it in VS2017 (I did not want to installl VS2015 only because of this one small project)

### CODE REVIEW

File: test.fsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
+﻿#I @"TickSpec/bin/Debug"
```
Comment (mchaloupka) [inline test.fsx]: We would definitely need to remove this file before merging into the official branch. But I am fine with it for now.

File: TickSpec/FeatureParser.fs [hunk 2]
```diff
@@ -52,7 +52,7 @@ let internal computeCombinations (tables:Table []):((string*string) list list) =
     values |> combinations
 
 /// Replace line with specified named values
-let internal replaceLine (xs:seq<string * string>) (scenario,n,tags,line,step) =
+let internal replaceLine (xs:seq<string * string>) (scenario,n,tags,(line:LineSource),step) =
```
Comment (mchaloupka) [inline TickSpec/FeatureParser.fs]: Why the type is needed here?

File: TickSpec/FeatureParser.fs [hunk 3]
```diff
@@ -80,24 +80,25 @@ let internal replaceLine (xs:seq<string * string>) (scenario,n,tags,line,step) =
...
+                |> Seq.map (fun table -> {Table=table; Tags=Array.empty})
+                |> Seq.toArray                    
+            scenarioName,scenarioTags,steps,Some(Array.append nonSharedExampleSets sharedExampleSets)
+    )
           
 /// Parses lines of feature
 let parseFeature (lines:string[]) =
     let toStep (_,_,_,line,step) = step,line
-    let featureName,background,scenarios,sharedExamples = parseBlocks lines     
+    let featureName,background,scenarios,sharedExampleSets = parseBlocks lines
```
Comment (mchaloupka) [inline TickSpec/FeatureParser.fs]: Is this renaming necessary? They are still shared examples. But I am fine if you prefer to rename them.

File: TickSpec/FeatureParser.fs [hunk 4]
```diff
@@ -80,24 +80,25 @@ let internal replaceLine (xs:seq<string * string>) (scenario,n,tags,line,step) =
...
-                scenarioName,tags,steps,Some(Array.append exampleTables sharedExamples)
-        )
+let internal appendSharedExamples sharedExampleSets scenarios  =
+    scenarios |> Seq.map (function 
+        | scenarioName,scenarioTags,steps,None ->
+            scenarioName,scenarioTags,steps,Some(sharedExampleSets)
+        | scenarioName,scenarioTags,steps,Some(exampleTables) ->
+            let nonSharedExampleSets = 
+                exampleTables
+                |> Seq.map (fun table -> {Table=table; Tags=Array.empty})
```
Comment (mchaloupka) [inline TickSpec/FeatureParser.fs]: Even nonSharedExamples may have tags, right?

**Comment by mchaloupka** [comment]
Ideally, the test.fsx should be removed. If you feel that you can use a better example then modify the added test for examples tagging.

File: TickSpec/BlockParser.fs [hunk 5]
```diff
@@ -156,23 +158,38 @@ let parseBlocks (featureLines:string[]) =
...
+                    |> Seq.choose identity
+                let examples = 
+                    examples 
+                    |> (function Some x -> Seq.append tables x | None -> tables)
+                (tag, examples, tags) |> Some
+        ) 
+    let sharedExamplesWithOfTag = 
+        shared 
+        |> Seq.choose (function
+            | Some tag,examples,tags -> Some(tag,examples)
```
Comment (michalkovy) [inline TickSpec/BlockParser.fs]: Looks like you are ignoring tags on shared example of tag, am I right? I would prefer having support for tags on all 3 types of examples - standard, shared and shared with of tag

File: TickSpec/FeatureParser.fs (hunk 2 above)
Comment (tencek) [reply TickSpec/FeatureParser.fs]: Without this annotation help, F# wrongly deduces the type of the `line` to be the `TaggedExample` (introduced by my change). Line 69 - referencing `line.Table` is why he thinks it is a `TaggedExample`. Alternative was to swap two blocks of code to reference `line.Bullets` first thus forcing F# to deduce the right type.

File: TickSpec/FeatureParser.fs (hunk 3 above)
Comment (tencek) [reply TickSpec/FeatureParser.fs]: It's just to have more consistent naming.

**Comment by tencek** [comment]
I agree the test.fsx is should be removed before we merge to the original repository. For now, I need it to make sure my changes works as intended.

File: TickSpec/BlockParser.fs (hunk 5 above)
Comment (tencek) [reply TickSpec/BlockParser.fs]: I will look at it.

File: TickSpec/FeatureParser.fs (hunk 4 above)
Comment (tencek) [reply TickSpec/FeatureParser.fs]: I will look at it.

Comment (tencek) [reply TickSpec/FeatureParser.fs]: For now, tags for non-shared examples seem to be unsupported.
When there appears some tag before the example table, it is assigned to the next block.
e.g:
```
Feature: Web login testing

    @scenariotag
    Scenario: Successful login using <browser>
    Given I use browser <browser>
    When I try to login using username <username> and password <password>
    Then I am logged in

    @nonsharedtag
    Examples:
    | username      | password      |
    | test          | psw           |
    | test2         | psw2          |

    Scenario: Show page using <browser>
    Given I use browser <browser>
    When I go to the main page
    Then The main page is displayed
```
the `@nonsharedtag` is assigned to the `Scenario: Show page using <browser>`

**Comment by mchaloupka** [comment]
Please handle the conflicts in the pull request.

**Comment by tencek** [comment]
Closing this pull request as I totally changed the approach to implement the TaggedExamples feature.

**Pull request closed without merging**